### PR TITLE
IB: Adding support for Neon

### PR DESCRIPTION
**Pull request by shamisp**
Signed-off-by: Pavel Shamis (Pasha) <pasharesearch@gmail.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1928/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2573/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2574/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1929/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@MattBBaker can you please check what is the story

**Comment by MattBBaker** [comment]
/dev/shm. The /dev/shm is already taking up half of the machine memory, I need to find some way to make these take an exclusive lock on /dev/shm.

**Comment by shamisp** [comment]
ok. false alarm

**Comment by shamisp** [comment]
@yosefe ?

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 1]
```diff
@@ -392,13 +392,19 @@ static UCS_F_ALWAYS_INLINE void
...
-#ifdef __SSE4_2__
+#if defined(__SSE4_2__)
     *(__m128i*)raddr = _mm_shuffle_epi8(
                 _mm_set_epi64x(rdma_rkey, rdma_raddr),
                 _mm_set_epi8(0, 0, 0, 0,            /* reserved */
                              8, 9, 10, 11,          /* rkey */
                              0, 1, 2, 3, 4, 5, 6, 7 /* rdma_raddr */
                              ));
+#elif defined(__ARM_NEON)
+    uint8x16_t table =  {7,6,5,4,3,2,1,0, /* rdma_raddr */
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: code style

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 2]
```diff
@@ -455,6 +461,19 @@ uct_ib_mlx5_set_ctrl_seg(struct mlx5_wqe_ctrl_seg* ctrl, uint16_t pi,
                                  4, 5,       /* sw_pi in BE */
                                  1           /* opmod */
                                  ));
+#elif defined(__ARM_NEON)
+    uint8x16_t table = {1,            /* opmod */
+                        5,4,          /* sw_pi in BE */
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: code style

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 3]
```diff
@@ -455,6 +461,19 @@ uct_ib_mlx5_set_ctrl_seg(struct mlx5_wqe_ctrl_seg* ctrl, uint16_t pi,
...
+                        2,            /* opcode */
+                        14,13,12,     /* QP num */
+                        8,            /* data size */
+                        16,           /* signature (set 0) */
+                        16,16,        /* reserved (set 0) */
+                        0,            /* signal/fence_mode */
+                        16,16,16,16}; /* immediate (set to 0)*/
+    uint32x4_t data = {(opcode << 16) | (opmod << 8) | (uint32_t)fm_ce_se,
+                        pi, ds, qp_num};
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: indentation

Comment (shamisp) [inline src/uct/ib/mlx5/ib_mlx5.h]: hmm it is 4 spaces ? what else ?

File: src/uct/ib/mlx5/ib_mlx5.h (hunk 2 above)
Comment (shamisp) [inline src/uct/ib/mlx5/ib_mlx5.h]: ?

File: src/uct/ib/mlx5/ib_mlx5.h (hunk 1 above)
Comment (shamisp) [inline src/uct/ib/mlx5/ib_mlx5.h]: ?

File: src/uct/ib/mlx5/ib_mlx5.h (hunk 3 above)
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: need to indent on column, 'pi' should be under '('

Comment (shamisp) [inline src/uct/ib/mlx5/ib_mlx5.h]: huh...wood have to adjust vim configuration

**Comment by shamisp** [comment]
@yosefe updated

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2587/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1943/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2588/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1945/ for details (Mellanox internal link).

**Pull request merged**